### DISCUSSION TITLE
Introduction of `BlockSchemes` to represent time-stepping schemes

### DISCUSSION THREAD

**Discussion by tlunet** [Parallel-in-Time/time4apint#16] [General]
Hi @MichelKrispin,

I've just introduced a new formulation for the time-integration schemes to be set by the user in the web api, see [this notebook](https://github.com/Parallel-in-Time/time4apint/blob/gaia/notebook/07_blockSchemes.ipynb) for a first quick introduction.

The idea is that you can directly retrieve from the `blockops` library all the implemented time-stepping schemes, along with their parameters, defaults values and documentation ... Normally, it should make your life easy when designing the generic component of the web interface to be used when setting the time-integration scheme.

:bell: note that I'll probably add some new time-stepping schemes later, so if you are generic enough when designing the web component, the new time-stepping scheme should be automatically included in the web api as soon as they are implemented in `blockops`, without modifying the website code :wink: 

The whole idea is based on the generic `blockops.utils.params` submodule, please take the time to look at it. In particular, it include some automatic testing for the parameter value from which you can probably benefit. And eventually, if you need anything more, this should be directly added there. 

Don't hesitate to write here if you have any request / comment / question / ...

**Comment by tlunet** [comment]
:bell: UPDATE

- the parameter approach with the `ParamClass` inheritance and `setParams` class wrapper is now applied to the `BlockProblem` class
- I've re-work the website documentation to start first with shorter and simpler application than the original one. For now @MichelKrispin you can focus on this [first one](https://github.com/Parallel-in-Time/time4apint/blob/gaia/doc/web-applications/1_accuracy.md)

**Comment by MichelKrispin** [comment]
Hey @tlunet ,

I've mirrored most of the parameter classes inside the web part to keep things separated and created converter functions to generate web parameters from blockops parameters.

It works quite well, but as a request, an additional math notation name for each parameter would be very nice as it can be displayed directly on the website. Maybe something like `lam=NumberParameter(math=r'\lambda')`, because right now, only the variable names are displayed.

And it might be more expressive to have a separate `Enumeration` type in the `blockops` library as well, as most of the `MultipleChoices` there only consist of enumerated string values.

And lastly, as a side note, the `docs.find(f'\n    {name} :')` line in `blockops/utils/params.py:101` that extracts the parameter documentation didn't work for me the first time as I had different indentations. But that wasn't too hard to fix so its not really a problem.

**Comment by tlunet** [comment]
Hey @MichelKrispin,

I've just added a `latexName` attribute for each parameter to be used for the website display (more explicit than `math` ...). It's automatically set using the name of the parameter when not explicitly given in `Parameter` constructor. Also, I've added a `uniqueID` property (with setter) that is automatically set to `ClassName_ParameterName`, and can be changed eventually by the website interface.
